File: src/spongify.ts
/**
 * Spongify a sentence
 * @param {string} sentence the input sentence
 * @returns {string} spongified sentence
 */
function spongify(sentence: string): string {
  return sentence
    .trim()
    .toLowerCase()
    .split('')
    .reduce((outputText, character): string => {
      const randomBoolean = getRandomBoolean();
      const spongeCharacter = randomBoolean
        ? character.toUpperCase()
        : character;

      return `${outputText}${spongeCharacter}`;
    }, '');
}

/**
 * Get a random boolean, used to determine
 * if character should be uppercase or not.
 * @returns {boolean}
 */
function getRandomBoolean(): boolean {
  const randomBinaryNumber = Math.floor(Math.random() * Math.floor(2));
  return Boolean(randomBinaryNumber);
}

export default spongify;

File: go/spongify.go
package main

import (
	"math/rand"
	"strings"
	"time"
)

// spongify a sentence
func spongify(input string) string {
	input = strings.TrimSpace(input)
	input = strings.ToLower(input)
	output := ""
	for _, character := range input {
		stringCharacter := string(character)
		randomBool := getRandomBoolean()
		if randomBool {
			output += strings.ToUpper(stringCharacter)
		} else {
			output += stringCharacter
		}
	}

	return output
}

// get a random boolean, used to determine
// if a character should be uppercase or not.
func getRandomBoolean() bool {
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	randomNumber := r.Intn(2)
	return randomNumber != 0
}
